Add tests for AND, ORR and TST. + Bugfix for SVC decoding + Compile target fix

dezuman opened the pull request.

dezuman committed:
add dummy tests for AND and ORR operators

dezuman committed:
Add dummy tests for TST operator

dezuman committed:
Add placeholder macro to step vm X times

vm.execute and SVC op breakpoint should be used here, but the op doesn't compile in the current toolchain. Thus this placeholder solution to avoid copy-pasting the vm.step command X times.

dezuman committed:
Add more test cases and improve comments

dezuman committed:
Refactor tests to use vm.execute() with SVC breakpoint

dezuman committed:
Add folder structure for organization

dezuman committed:
Add one-step script to build and run assembly tests

dezuman committed:
Remove unnecessary test cases

dezuman committed:
Correct and clarify test descriptions

dezuman committed:
Add note on test reference and future TODO

Earlz commented on tests/common.rs:
```diff
@@ -22,6 +22,18 @@ pub fn create_test_vm(file: &str) -> NarmVM {
+// Macro to step VM several times
+// Will be replaced with vm.execute and SVC breaks when the SVC-op-not-compiling issue is fixed
+#[macro_export]
+macro_rules! multistep_vm {
```
This isn't needed now right?

Earlz commented on tests/test_and/test_and_flag_neg.s:
```diff
@@ -0,0 +1,13 @@
+    ands r0, r0
+    svc #0x1
+1: 
+    b 1b @ Halt
```
This can likely be removed safely since it is unreachable code

Earlz commented on tests/test_and/test_and_flag_neg.s:
```diff
@@ -0,0 +1,13 @@
+    lsls r0, #0xf
+    lsls r0, #0xf
+    ands r0, r0
+    svc #0x1
```
Please use `svc #0xFF` to halt tests. This is the svc call which will be used in the production version of Neutron so might as well match it here.

Earlz commented on tests/test_tst.rs:
```diff
@@ -0,0 +1,37 @@
+    let mut vm = create_test_vm("test_tst_flag_neg");
+    vm.execute().unwrap();
+    vm.print_diagnostics();
+    assert!(vm.cpsr.n);
```
There should be an additional test/assert that confirms that the registers specified are not modified, like would be done for the AND instruction

Earlz replied:
Also a test that multiple arguments (ie, more than just r0) works as expected

Earlz commented on tests/test_and.rs:
```diff
@@ -0,0 +1,47 @@
+    let mut vm = create_test_vm("test_and_flag_neg");
+    vm.execute().unwrap();
+    vm.print_diagnostics();
+    assert!(vm.cpsr.n);
```
For all of these tests, the C (carry) flag shouldn't be affected. Write a test/assert to confirm C is not modified

dezuman replied:
Is it good enough that I just check that none of the cases set C, or should I also set the flag and check that it isn't unset?

dezuman replied:
I'll just check to make sure they aren't altered by existing tests for now

Earlz commented on tests/build.sh:
```diff
@@ -2,14 +2,14 @@
   OBJECTFILE="bin/temp/$(basename "$f" .s).o"
   FINALFILE="bin/$(basename "$f" .s)"
-  arm-none-eabi-as -march=armv6-m -o $OBJECTFILE $f
+  arm-none-eabi-as -march=armv6s-m -o $OBJECTFILE $f
```
ACK (for public knowledge, armv6-m targeting for some reason disallows the SVC instruction. Most Cortex-M0 etc projects use armvs-m to workaround this limitation)

dezuman replied:
Added a code comment about the reason for the switch

Earlz reviewed (changes requested):
Some fairly minor changes requested around a few additional behaviors to test and slight style changes

dezuman committed:
Remove unreachable assembly code

dezuman committed:
Standardize svc op to "svc #0xFF"

dezuman committed:
Add checks that flags that shouldn't be altered aren't

dezuman committed:
Update TST test to check that target reg is unaltered

dezuman committed:
Updated TST test to use several registers for broader checking

dezuman committed:
Add macro to check all 8 low registers

dezuman committed:
Refactor all tests to use new reg assert macro

Even the ones that didn't check registers before, so the tests are broader now

dezuman committed:
Add macro to check all 4 condition flags

dezuman committed:
Refactor all tests to use new flag assert macro

All tests now assert value of all 4 condition flags, so the tests are broader

dezuman committed:
Standardize and clean up all tests

dezuman committed:
Refactoring and cleanup after rebase

Assembly code is now inline in the rust files!
*everyone likes that*

dezuman committed:
Add structured way to neatly assert the VMs collected state

dezuman committed:
Refactor tests for new format

dezuman committed:
Refactor state struct to a nicer format + misc reformat

Included: 
- Tests adapted to new format
- Standardization of static hex declaration to "0x00_00_00_00" format that works Rust and Asm
- Binary representation in comments maybe finally standardized

Earlz commented on test.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
This should no longer be needed

Earlz commented:
For reference `movs r0, #0x00_00_00_0F` the formatting on this is not strictly required, but should be used where clarity is needed for bigger values. In the specific small number cases here, 0x0F would've been fine

dezuman committed:
Add changes missing from previous commit

dezuman committed:
Add a small description to the VM state assertion infrastructure

dezuman committed:
Bestove final death on accidental rebase zombie file
